R/install_reqs.R

```diff
@@ -45,31 +34,46 @@ install_unversioned = function(elem, installed, dryrun, verbose, repo){
   #' @return the number of failures
   failures = 0
 
-  if (length(elem) > 0) {
-    for (i in elem) {
-      if (is.na(installed[i])) {
-        available_versions = get_available_versions(i)
-        vmess(sprintf(NOTE_PACKAGE_NOT_INSTALLED, i), verbose)
-        version = tail(sort(available_versions), 1)
-        vmess(sprintf(NOTE_INSTALL_PACKAGE, i, sprintf(NOTE_INSTALL_PACKAGE_VERSION, version)), verbose)
-        if (!dryrun) {
-          tryCatch(install.packages(i, repos = repo),
-                   error = function(x){
-                     failures <<- failures + 1
-                     message(sprintf(ERROR_OTHER_FAILURE, i))
-                   })
-        }
-      }
+  for (i in elem) {
+    available_versions = get_available_versions(i)
+    if (is.null(available_versions)){
+      failures = failures + 1
+      vmess(sprintf(ERROR_NO_PACKAGE_EXISTS, i), TRUE)
+    } else if (is.na(installed[i])) {
+      vmess(sprintf(NOTE_PACKAGE_NOT_INSTALLED, i), verbose)
+      version = tail(sort(available_versions), 1)
+      vmess(sprintf(NOTE_INSTALL_PACKAGE, i, sprintf(NOTE_INSTALL_PACKAGE_VERSION, version)), verbose)
+      failures = failures + install_cran_package(package = i,
+                                                 version = NULL,
+                                                 repo = repo,
+                                                 dryrun = dryrun)
     }
   }
 
   return(failures)
 }
 
+install_cran_package = function(package, version, repo, dryrun){
+  failures = 0
+  if (!dryrun) {
+    tryCatch(remotes::install_version(package,
+                                       version = version,
+                                       repos = repo,
+                                       quiet = TRUE),
+             error = function(x){
+               failures <<- failures + 1
+               message(sprintf(ERROR_OTHER_FAILURE, package))
+             })
+  }
+  return(failures)
+}
+
 process_versioned_requirement = function(req){
-  comp = sub(paste0("^.*?(", paste(COMPS, collapse = "|"), ").*$"), "\\1", req)
+  comps = paste0("=|", paste(COMPS, collapse = "|"))
+  pattern = "^.*?(=|%s).*$"
+  comp = sub(sprintf(pattern, comps), "\\1", req)
   if (comp == "=") comp = COMP_EXACTLY_EQUAL
```

Comment:
I think this processing could be improved from a readability standpoint.  I don't know the best way to approach this processing while maintaining high readability, but I think it's worth exploring.

Maybe something that relies on a fixed pattern using the `COMPS` and then fall back to checking for `=`?

(The below is not tested extensively, and I'm fully in love with it)

```r
> vapply(COMPS, grepl, x='mgsub==4', fixed=TRUE, FUN.VAL=logical(1))
   ==    <=     <    >=     >    ~=    != 
 TRUE FALSE FALSE FALSE FALSE FALSE FALSE 
```

Reply:
The above runs into issues with multiple matches when using something like `>=`.

```r
> vapply(COMPS, grepl, x='mgsub>=4', fixed=TRUE, FUN.VAL=logical(1))
   ==    <=     <    >=     >    ~=    != 
FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE 
```

Reply:
the COMPS could be sorted by nchar in the above to fix the issue, but i could just be trying to fix a solution that is subpar from the start